telemetry/internal/appinsights/inmemorychannel.go

```diff
@@ -0,0 +1,311 @@
+	channel.start()
+}
+
+// Part of channel accept loop: Initialize buffer and accept first message, handle controls.
+func (channel *inMemoryChannel) start() {
+	items := make([]batchItem, 0, channel.batchSize)
+	timer := time.NewTimer(time.Hour)
+	timer.Stop() // Stop timer until we need it.
+	var dropped int
+	for {
+		select {
+		case item := <-channel.collectChan:
+			if item == nil {
+				panic("received nil event")
+			}
+			if channel.throttled.Load() {
+				// Check if there is space to add the item to the batch.
+				// If not, then drop the event.
+				if len(items) < channel.batchSize {
+					items = append(items, batchItem{item, 0})
+				} else {
+					dropped++
+				}
+				continue
+			}
+			items = append(items, batchItem{item, 0})
+			if len(items) >= channel.batchSize {
+				timer.Stop()
+				channel.sendBatch(items)
+				items = items[:0]
+			} else if len(items) == 1 {
+				// Start the timer if this is the first item in the batch.
+				timer.Reset(channel.batchInterval)
+			}
+
+		case <-channel.flushChan:
+			if channel.throttled.Load() {
+				// Ignore the flush request if we are throttled.
+				continue
+			}
+			timer.Stop()
+			channel.sendBatch(items)
+			items = items[:0]
+
+		case <-timer.C:
+			if channel.throttled.Load() {
+				// When throttled, the timer is reset to the retry time,
+				// so if we get here, then we're no longer throttled.
+				channel.throttled.Store(false)
+				if dropped > 0 {
+					channel.logf("Dropped %d telemetry items due to throttling", dropped)
+					dropped = 0
+				}
+			}
+			channel.sendBatch(items)
+			items = items[:0]
+
+		case msg := <-channel.retryChan:
+			// If there is not enough space in the batch, drop the items.
+			space := channel.batchSize - len(items)
+			if space < len(msg.items) {
+				dropped += len(msg.items) - space
+				msg.items = msg.items[:space]
+			}
+			items = append(items, msg.items...)
+
+			if msg.throttled {
+				channel.throttled.Store(true)
+			}
+			if msg.retryAfter.IsZero() {
+				// If the retry time is not set, use the default batch interval.
+				timer.Reset(channel.batchInterval)
+			} else {
+				timer.Reset(time.Until(msg.retryAfter))
+			}
+
+		case <-channel.cancelCtx.Done():
+			// This is the only path to exit the loop.
+			timer.Stop()
+			return
+		}
+	}
+}
+
+var itemsBuf = sync.Pool{
+	New: func() any {
+		buf := make([]batchItem, 0, 1024)
+		return &buf
+	},
+}
+
+// sendBatch schedules a batch of items for transmission.
+func (channel *inMemoryChannel) sendBatch(items []batchItem) {
+	channel.sendQueueMu.Lock()
+	defer channel.sendQueueMu.Unlock()
+
+	if len(items) == 0 {
+		// Cancel the accept loop if we are closed and the queue is empty.
+		if len(channel.sendQueue) == 0 && channel.closed.Load() {
+			channel.cancelCauseFunc(errClosed)
+		}
+		return
+	}
+
+	// Copy the items to a temporary buffer to let the caller
+	// reuse the item slice.
+	buf := itemsBuf.Get().(*[]batchItem)
+	*buf = (*buf)[:0]
+	*buf = append(*buf, items...)
```

Comment:
I'm vaguely aware of [a concern with `sync.Pool` usage for buffers](https://github.com/golang/go/issues/23199) where if you sometimes grow the buffers, you might end up accumulating lots of buffers that are all needlessly high-capacity, putting pressure on memory. Currently each of these bufs have 1024 cap, and if `items` is longer than that, it looks like this buf would grow, running into that gotcha.

I don't know for sure if that concern applies here or not, but if so, perhaps we should e.g. grab multiple bufs from the pool if `len(items)` > 1024 rather than growing the buf we happen to get?

Reply:
Good point. The `items` size is capped to `Client.MaxBatchSize` (defaults to 1024), so the polled buffer sizes can't grow unbounded and cause an OOM, which is the main complain in that issue. I've added a comment to note this limit, and also moved the pool from being a global variable to being a property of the channel so that we can allocate buffers knowing for sure the maximum capacity.